### ISSUE TITLE
Vertically-centered text is too low

### ISSUE THREAD

**Issue by viridia** [bevyengine/bevy#15971]
## Bevy version

0.15.0-dev, although this bug has been around a long time.

## What you did

In a flexbox or grid, place a text element with vertical centering.

## What went wrong

The text appears 1 or 2 pixels lower than it should, as seen in the following example screenshot:

![Image](https://github.com/user-attachments/assets/b60db204-59ea-4768-be62-326952203f0a)

## Additional information

I had hoped that the migration to cosmic-text would fix this, but apparently not.

@UkoeHB

**Comment by UkoeHB** [comment]
Is the text itself not centered, or the text node not centered? I assume you're seeing the problem with `JustifyContent::Center`?

**Comment by viridia** [comment]
This particular example is using align, not justify (it's a flex row, not a column) - but I think the type of alignment doesn't matter. I've seen it in a bunch of different situations. I suspect it's a problem in the way the text is being measured, or perhaps a rounding issue.

**Comment by ickshonpe** [comment]
This isn't a layout problem but the text not centered I think, though when I've seen it was too high, not too low. You can tell for certain by setting a background color on the text itself.

**Comment by viridia** [comment]
I'm using OpenSans.

**Comment by UkoeHB** [comment]
I also had problems with OpenSans and switched to FiraSans.

**Comment by viridia** [comment]
Here's what it looks like with Ubuntu:
![Image](https://github.com/user-attachments/assets/24fcf4a0-ce43-4a69-8cbc-226c029741a5)

**Comment by ickshonpe** [comment]
This makes it clear, definitely malpositioned:
![Image](https://github.com/user-attachments/assets/ad56d493-5554-4bc3-bfe4-0ddb811f038a)

**Comment by viridia** [comment]
It does look better with FiraSans:

![Image](https://github.com/user-attachments/assets/4af7f45f-2cd1-47a8-93c6-41d2140b8d82)

**Comment by mamekoro** [comment]
Does anyone know if there is an API in Bevy or its dependencies that helps visualize how text is laid out that would be helpful in investigating this type of problem? (Something like `Paint.FontMetrics` or `Paint.getTextBounds()` in Android).

**Comment by ickshonpe** [comment]
I thought it could be some problem with my layout rounding hacks but there's no difference after removing them.

**Comment by mamekoro** [comment]
I'm not sure if it is really malpositioned. I opened OpenSans-Regular.ttf on FontForge and it seems that glyph y originally reached below the descent line.

![Image](https://github.com/user-attachments/assets/dce7110b-ae1c-4e1f-8f06-c4d13e852d74)

**Comment by doup** [comment]
This is what I see in Chrome:

- Open Sans Regular
- **First pink box**: `font-size: 60px, line-height: 0.5 (30px)`
- **Second pink box**: `font-size: 60px, line-height: 1 (60px)`
- **Rectangle**: `60x60px`

![Image](https://github.com/user-attachments/assets/1f4aa657-72de-4950-b54c-b9b27377c8ae)

I think it defines the text box based on the ascenders height (not the uppercase) and the baseline (red lines), then centers vertically the text based on the line-height (pink boxes).  It can overflow in both directions.

![Image](https://github.com/user-attachments/assets/2e841de3-685b-4e49-a7d2-a0dff65210d4)

I just realized I don't understand what `font-size: 60px` means. I knew different fonts have different cap-height/x-height… but what's 60px here? 😅

**Comment by BenjaminBrienen** [comment]
the number of pixels in height

**Comment by doup** [comment]
@BenjaminBrienen yes, but height of what? Because it's not of the caps, or from descenders to ascenders…

I forgot to answer myself, **`font-size` refers to the height of the ["em square"](http://designwithfontforge.com/en-US/The_EM_Square.html)** as it's noted in the [web specs](https://www.w3.org/TR/CSS2/fonts.html#font-size-props):
> The font size corresponds to the em square, a concept used in typography. Note that certain glyphs may bleed outside their em squares.

What's interesting is the note, _"glyphs may bleed outside"_… and I suppose that's what I see in my screenshot. The em-square is 60px, but some diacritics (`Ñ`) or descenders (`y`) bleed out. I tried to check this in FontForge but I have no idea how I can visualize it.

![Image](https://github.com/user-attachments/assets/7e8693e2-70d2-4f49-a0bf-19a80e461536)


**Which leads me to think that either Bevy or cosmic-text it's not considering the "em-square"?** Why? It's just a hunch after seeing the Firefox/Chrome screenshot, which clearly bleeds from both top and bottom which suggest that "Open Sans Regular" em square bleeds? ickshonpe's screenshots have way too much space both at the top and bottom, in one of the screenshots even the `y` doesn't overflow. (btw… why the inconsistency? maybe the screenshots have different line-height?)

![Image](https://github.com/user-attachments/assets/50527e12-d895-4af4-ac7e-45c50098bbc5)

**Comment by doup** [comment]
Ignore my previous comments (🙈), I just realized that Bevy uses 1.2 hard-coded line-height, which gives the same result as in Firefox:

![Image](https://github.com/user-attachments/assets/c576175b-2567-41bf-a375-5ff13d34e7dd)

**Comment by doup** [comment]
For what is worth… `bevy-reactor` vs Firefox: **1px offset**.

![Image](https://github.com/user-attachments/assets/a36360a3-4b71-46d8-87ea-ad1419f37d44)

**Comment by BenjaminBrienen** [comment]
But which is correct? 🤔

**Comment by doup** [comment]
IMO, OP is correct, text looks too low in Bevy. `Md`/`Xs`/`Xxs` look particularly bad in the first screenshot, or the first two rows (which are `Md` size).

Also note that between the two screenshots there is some anti-aliasing difference. But no idea if that's relevant. It could be because FF uses another anti-aliasing? Or the FF text is not pixel grid aligned and it's 0.5px higher or whatever? 🤷‍♂

**Comment by musjj** [comment]
I had the same issue, but the opposite (the text is too high): https://github.com/bevyengine/bevy/issues/16627

My solution was to edit the font using FontForge, go to `Element` -> `Font Info...` -> `OS/2` -> `Metrics` and increase the `Win Ascent`, `Typo Ascent` and `HHead Ascent`. Not sure if there's any nasty hidden side-effects, but it works as a quick hack.

**Comment by viridia** [comment]
Here's an updated screenshot showing the issue with the latest main (with Parley):

<img width="230" height="43" alt="Image" src="https://github.com/user-attachments/assets/9484dd92-4cdc-4808-9d8b-7f66b3662cdd" />

This was taken on MacOS using this code:

```rust
fn button_caption(text: &'static str) -> impl Scene {
    bsn! {
        Text(text)
        TextColor(palettes::css::BLACK)
        TextFont {
            font: FontSourceTemplate::SansSerif,
            font_size: FontSize::Px(20.0)
            weight: FontWeight::BOLD,
        }
    }
}
```

@ickshonpe

**Comment by ickshonpe** [comment]
Maybe it's some sort of accumulated rounding error, try adding `LayoutConfig { use_rounding: false }` to the root node?
